### PR TITLE
RDoc: Enhanced introduction for Enumerable

### PR DESCRIPTION
**Pull request by BurdetteLamar**
I have enhancements for the 49 methods, too, but will put up piecemeal to facilitate review.

### CODE REVIEW

File: enum.c [hunk 1]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ * #slice_after::    Returns a new \Enumerator whose entries are a partition of +self+,
+                     based either on a given +object+ or a given block.
+ * #slice_before::   Returns a new \Enumerator whose entries are a partition of +self+,
+                     based either on a given +object+ or a given block.
+ * #slice_when::     Returns a new \Enumerator whose entries are a partition of +self+
+                     based on the given block.
+ * #chunk::          Returns elements organized into chunks as specified by the given block.
+ * #chunk_while::    Returns elements organized into chunks as specified by the given block.
+ *
+ * === Methods for Searching
```
Comment (baweaver) [inline enum.c:4239]: Perhaps "Methods for Searching and Filtering"?

File: enum.c [hunk 2]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ *
+ * These methods return elements that meet a specified criterion.
+ *
+ * #find, #detect::              Returns an element selected by the block.
+ * #find_all, #filter, #select:: Returns elements selected by the block.
```
Comment (baweaver) [inline enum.c:4244]: "selected by the block" may be unfamiliar, perhaps something like this?:

```
Returns all elements for which the block returns +true+
```

File: enum.c [hunk 3]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ * #each_entry::       Calls the block with each successive element
+ *                     (slightly different from #each).
+ * #each_with_index::  Calls the block with each successive element and its index.
+ * #each_with_object:: Calls the block with each successive element and a given object.
+ * #each_slice::       Calls the block with successive slices of elements.
+ * #each_cons::        Calls the block with successive slices of elements
+ *                     (different from #each_slice).
+ * #reverse_each::     Calls the block with each successive element, in reverse order.
+ *
+ * === Methods for Filtering
```
Comment (baweaver) [inline enum.c:4269]: These methods feel more like transformations rather than filtering a list.

File: enum.c [hunk 4]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ *
+ * #map, #collect::             Returns objects returned by the block.
```
Comment (baweaver) [inline enum.c:4271]: Referencing transformations might make this clearer:

```
Returns a new collection transformed by the block
```

Not sure what type of wording would work best there.

File: enum.c [hunk 5]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ *
+ * === Methods for Iterating
+ *
+ * #each_entry::       Calls the block with each successive element
+ *                     (slightly different from #each).
+ * #each_with_index::  Calls the block with each successive element and its index.
+ * #each_with_object:: Calls the block with each successive element and a given object.
+ * #each_slice::       Calls the block with successive slices of elements.
+ * #each_cons::        Calls the block with successive slices of elements
+ *                     (different from #each_slice).
```
Comment (baweaver) [inline enum.c:4266]: This one will be hard to describe in one sentence, but using the same as `each_slice` might be confusing here.

**Comment by baweaver** [comment]
Definitely a fan of adding an intro section of "what to use" type documentation here, as Enumerable is so commonly used.

**Comment by BurdetteLamar** [comment]
@baweaver, I'm pretty well satisfied that this PR is an improvement.  I'd like to merge as is, and move on to the methods doc.  Then you can PR any changes you'd like to see.

**Comment by baweaver** [comment]
I may not understand the way that PRs work with Ruby itself, but would it not be best to hash out wording before merging rather than merge a potentially confusing patch? I could certainly raise PRs afterwards.

File: enum.c (hunk 5 above)
Comment (jeremyevans) [reply enum.c:4266]: `each_slice`: successive non-overlapping slices
`each_cons`: successive overlapping slices

File: enum.c (hunk 3 above)
Comment (jeremyevans) [reply enum.c:4269]: At least `cycle` here is not a transformation.  I would just say "Other Methods".

File: enum.c (hunk 1 above)
Comment (BurdetteLamar) [reply enum.c:4239]: Done.

File: enum.c (hunk 2 above)
Comment (BurdetteLamar) [reply enum.c:4244]: If the reader doesn't know #select and #reject, the meanings fall through to regular English, which is what I meant in the first place.  These 1-liners I've meant to be indicative, not definitive.  Any questions can be resolved by following the links.

File: enum.c (hunk 4 above)
Comment (BurdetteLamar) [reply enum.c:4271]: I'm content with stet.  The returns may not be transforms in any meaningful way.  Could, for example, be just +true+ and +false+.

File: enum.c (hunk 5 above)
Comment (BurdetteLamar) [reply enum.c:4266]: Bravo, JE.  Done.

**Comment by BurdetteLamar** [comment]
@baweaver, my re-request thingy does not work, so pinging you here.

File: enum.c (hunk 3 above)
Comment (BurdetteLamar) [reply enum.c]: "Other Methods" it is.

File: enum.c (hunk 2 above)
Comment (marcandre) [reply enum.c:4244]: Sounds ok to me.

File: enum.c [hunk 6]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ * #chunk_while::    Returns elements organized into chunks as specified by the given block.
+ *
+ * === Methods for Searching and Filtering
+ *
+ * These methods return elements that meet a specified criterion.
+ *
+ * #find, #detect::              Returns an element selected by the block.
+ * #find_all, #filter, #select:: Returns elements selected by the block.
+ * #grep::                       Returns elements selected by a given object or block.
+ * #grep_v::                     Returns elements selected by a given object or block.
```
Comment (marcandre) [inline enum.c:4246]: These are harder to summarize accurately, as they can also map the elements (and thus no necessarily "returns elements")

File: enum.c [hunk 7]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ * To use module \Enumerable in a collection class:
+ * - Include it:
+ *     include Enumerable
+ * - Implement method <tt>#each</tt>
+ *   which must yield successive elements of the collection.
+ *   This method will be called by almost any \Enumerable method.
+ * - Implement method <tt><=></tt> (or just inherit
+ *   {Object#<=>}[https://ruby-doc.org/core/Object.html#method-i-3C-3D-3E]).
+ *   The method will be called by many \Enumerable methods, including
+ *   <tt>#max</tt>, <tt>#min</tt>, and <tt>#sort</tt>.
```
Comment (marcandre) [inline enum.c:4291]: The method `<=>` need be implemented on the *elements*, not on the *collection*. Moreover, implementing it is not sufficient, as returning `nil` (like `Object#<=>`) will still raise.
I wouldn't include this paragraph here at all.

File: enum.c [hunk 8]
```diff
@@ -4179,13 +4179,137 @@ enum_uniq(VALUE obj)
...
+ *
+ * == \Enumerable in Ruby Core Classes
+ * Some Ruby classes include module \Enumerable:
```
Comment (marcandre) [inline enum.c:4294]: "include *the* module `Enumerable`", or "include `Enumerable`"

File: enum.c (hunk 6 above)
Comment (BurdetteLamar) [reply enum.c]: Moved and slightly revised.

File: enum.c (hunk 7 above)
Comment (BurdetteLamar) [reply enum.c]: Deleted, and thanks @marcandre for preventing this blunder.

File: enum.c (hunk 8 above)
Comment (BurdetteLamar) [reply enum.c]: Fixed. "include Enumerable"

**Comment by marcandre** [comment]
Oops, I thought of doing a squash merge, but forgot to update the PR's message.

Thanks @BurdetteLamar 👍

**Pull request merged**